client/Wayland/wlf_cliprdr.c

```diff
@@ -0,0 +1,863 @@
+	format->formatId = formatId;
+	format->formatName = NULL;
+	if (name && (formatId >= CF_MAX))
+		format->formatName = _strdup(name);
+}
+
+static void wlf_cliprdr_add_client_format(wfClipboard* clipboard, const char* mime)
+{
+	if (wlf_mime_is_html(mime))
+	{
+		UINT32 formatId = ClipboardGetFormatId(clipboard->system, "HTML Format");
+		wfl_cliprdr_add_client_format_id(clipboard, formatId);
+	}
+	else if (wlf_mime_is_text(mime))
+	{
+		wfl_cliprdr_add_client_format_id(clipboard, CF_TEXT);
+		wfl_cliprdr_add_client_format_id(clipboard, CF_OEMTEXT);
+		wfl_cliprdr_add_client_format_id(clipboard, CF_UNICODETEXT);
+	}
+	else if (wlf_mime_is_image(mime))
+	{
+		UINT32 formatId = ClipboardGetFormatId(clipboard->system, "image/bmp");
+		wfl_cliprdr_add_client_format_id(clipboard, formatId);
+		wfl_cliprdr_add_client_format_id(clipboard, CF_DIB);
+	}
+	wlf_cliprdr_send_client_format_list(clipboard);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_send_data_request(wfClipboard* clipboard,
+        UINT32 formatId)
+{
+	CLIPRDR_FORMAT_DATA_REQUEST request;
+	ZeroMemory(&request, sizeof(CLIPRDR_FORMAT_DATA_REQUEST));
+	request.requestedFormatId = formatId;
+	return clipboard->context->ClientFormatDataRequest(clipboard->context,
+	        &request);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_send_data_response(wfClipboard* clipboard, const BYTE* data,
+        size_t size)
+{
+	CLIPRDR_FORMAT_DATA_RESPONSE response;
+	ZeroMemory(&response, sizeof(CLIPRDR_FORMAT_DATA_RESPONSE));
+	response.msgFlags = (data) ? CB_RESPONSE_OK : CB_RESPONSE_FAIL;
+	response.dataLen = size;
+	response.requestedFormatData = data;
+	return clipboard->context->ClientFormatDataResponse(clipboard->context,
+	        &response);
+}
+
+BOOL wlf_cliprdr_handle_event(wfClipboard* clipboard, const UwacClipboardEvent* event)
+{
+	if (!clipboard || !event)
+		return FALSE;
+
+	if (!clipboard)
+		return FALSE;
+
+	switch (event->type)
+	{
+		case UWAC_EVENT_CLIPBOARD_AVAILABLE:
+			clipboard->seat = event->seat;
+			return TRUE;
+
+		case UWAC_EVENT_CLIPBOARD_OFFER:
+			WLog_Print(clipboard->log, WLOG_INFO, "client announces mime %s", event->mime);
+			wlf_cliprdr_add_client_format(clipboard, event->mime);
+			return TRUE;
+
+		case UWAC_EVENT_CLIPBOARD_SELECT:
+			WLog_Print(clipboard->log, WLOG_DEBUG, "client announces new data");
+			wlf_cliprdr_free_client_formats(clipboard);
+			return TRUE;
+
+		default:
+			return FALSE;
+	}
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_send_client_capabilities(wfClipboard* clipboard)
+{
+	CLIPRDR_CAPABILITIES capabilities;
+	CLIPRDR_GENERAL_CAPABILITY_SET generalCapabilitySet;
+	capabilities.cCapabilitiesSets = 1;
+	capabilities.capabilitySets = (CLIPRDR_CAPABILITY_SET*) &
```

Comment:
just nitpicking but it would be cool to adjust the styling here and few line below (line 319)

Reply:
I'd prefer #4350 finally being agreed upon and merged ;)